storage/txcache/txListForSender.go

```diff
@@ -0,0 +1,193 @@
+		items:       list.New(),
+		orderNumber: globalIndex,
+		sender:      sender,
+	}
+}
+
+// AddTx adds a transaction in sender's list
+// This is a "sorted" insert
+func (listForSender *txListForSender) AddTx(txHash []byte, tx data.TransactionHandler) {
+	// We don't allow concurent interceptor goroutines to mutate a given sender's list
+	listForSender.mutex.Lock()
+
+	nonce := tx.GetNonce()
+	mark := listForSender.findTxWithLargerNonce(nonce)
+	newNode := txListForSenderNode{txHash, tx}
+
+	if mark == nil {
+		listForSender.items.PushBack(newNode)
+	} else {
+		listForSender.items.InsertBefore(newNode, mark)
+	}
+
+	listForSender.mutex.Unlock()
+}
+
+func (listForSender *txListForSender) findTxWithLargerNonce(nonce uint64) *list.Element {
+	for element := listForSender.items.Front(); element != nil; element = element.Next() {
+		value := element.Value.(txListForSenderNode)
+		if value.tx.GetNonce() > nonce {
+			return element
+		}
+	}
+
+	return nil
+}
+
+// RemoveTx removes a transaction from the sender's list
+func (listForSender *txListForSender) RemoveTx(tx data.TransactionHandler) bool {
+	// We don't allow concurent interceptor goroutines to mutate a given sender's list
+	listForSender.mutex.Lock()
+
+	marker := listForSender.findTx(tx)
+	isFound := marker != nil
+	if isFound {
+		listForSender.items.Remove(marker)
+	}
+
+	listForSender.mutex.Unlock()
+
+	return isFound
+}
+
+// RemoveHighNonceTxs removes "count" transactions from the back of the list
+func (listForSender *txListForSender) RemoveHighNonceTxs(count uint32) [][]byte {
+	removedTxHashes := make([][]byte, count)
+
+	listForSender.mutex.Lock()
+
+	index := uint32(0)
+	var previous *list.Element
+	for element := listForSender.items.Back(); element != nil && count > index; element = previous {
+		// Remove node
+		previous = element.Prev()
+		listForSender.items.Remove(element)
+
+		// Keep track of removed transaction
+		value := element.Value.(txListForSenderNode)
+		removedTxHashes[index] = value.txHash
+
+		index++
+	}
+
+	listForSender.mutex.Unlock()
+
+	return removedTxHashes
+}
+
+func (listForSender *txListForSender) findTx(txToFind data.TransactionHandler) *list.Element {
+	for element := listForSender.items.Front(); element != nil; element = element.Next() {
+		value := element.Value.(txListForSenderNode)
+		if value.tx == txToFind {
+			return element
+		}
+	}
+
+	return nil
+}
+
+// HasMoreThan checks whether the list has more items than specified
+func (listForSender *txListForSender) HasMoreThan(count uint32) bool {
+	return uint32(listForSender.items.Len()) > count
+}
+
+// IsEmpty checks whether the list is empty
+func (listForSender *txListForSender) IsEmpty() bool {
+	return listForSender.items.Len() == 0
+}
+
+// startBatchCopying resets the internal state used for copy operations
+func (listForSender *txListForSender) startBatchCopying(batchSize int) {
+	// We cannot copy or start copy from multiple goroutines at the same time
+	listForSender.mutex.Lock()
```

Comment:
because between startBatchCopying call and copyBatchTo call, a removal could be triggered on a go routine (e.g RemoveTx is waiting on the listForSender.mutex because it was taken first by startBatchCopying, and then after startBatchCopying releases the mutex it is immediately taken by RemoveTx) there is a chance that you remove the front item. 
Not sure what inconsistency this could cause afterwards, so I think startBatchCopying should be under the same critical section as copyBatchTo, with no release of mutex in the middle.

Reply:
Fixed, and `startBatchCopying` and `copyBatchTo` are now a single function.